### PR TITLE
Swap to Vec instead of BTreeMap in Irep to reduce memory usage

### PR DESCRIPTION
**Pull request by tedinski**
### Description of changes: 

When running a std lib build test with `time -v` I went from

```
    Maximum resident set size (kbytes): 11866204
    Maximum resident set size (kbytes): 11866052
```

to

```
    Maximum resident set size (kbytes): 9417760
    Maximum resident set size (kbytes): 9417552
```

Which is a savings of 2.4 GB of memory. This isn't enough to unblock running this test in CI yet, but it is (a) I think a good change and (b) doesn't get in the way of the real fix Celina is working on.

(The `Json` type we use from `rustc_serialize` also uses `BTreeMap` and it likely accounts for another good 2.5GB (at least, possibly even 5GB) of wasted memory. But that's part of rustc, so let's not change that. So our savings here is despite still having to build all those BTreeMaps.)

### Resolved issues:

Part of #608. Possibly all of it, if we determine string interning there isn't the win I thought it might be.

### Call-outs:

1. We don't actually call `lookup` anywhere, it's only used in `from_irep` which we don't use in RMC normally. So a small part of this code is untested afaict.
2. There's a slight change in theoretical behavior in `with_named_sub`. If the entry was already present it would previously get updated, while now we'd "duplicate" it. I did inspect all uses of this function, and we don't have any place where we're trying to do an "update" with this function, so there's no issue right now, but that's possibly a future footgun?
3. I think we should merge only if we think this is a good representation change. I think it is, but the memory savings won't be as important (I think) once the streaming-serialization fix is in.

### Testing:

* How is this change tested? existing regression suite

* Is this a refactor change? yes

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/cbmc/src/irep/to_json.rs [hunk 1]
```diff
@@ -11,7 +12,11 @@ impl ToJson for Irep {
             output.insert("sub".to_string(), self.sub.to_json());
         }
         if !self.named_sub.is_empty() {
-            output.insert("namedSub".to_string(), self.named_sub.to_json());
+            let mut d = BTreeMap::new();
```
Comment (danielsn) [inline compiler/cbmc/src/irep/to_json.rs:15]: You can use `BTreeMap::FromIter()` here to avoid the loop

**Comment by danielsn** [comment]
https://github.com/model-checking/rmc/pull/620 does the same thing, using a `vector_map` crate.

File: compiler/cbmc/src/irep/to_json.rs (hunk 1 above)
Comment (tedinski) [reply compiler/cbmc/src/irep/to_json.rs:15]: Yeah, that required a map though so it seemed like more complicated code

**Comment by tedinski** [comment]
closing in favor of #620

**Pull request closed without merging**